components/action-mgt/org.wso2.carbon.identity.action.execution/src/test/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImplTest.java

```diff
@@ -304,6 +305,49 @@ public void testExecuteSuccess() throws Exception {
         assertEquals(actualStatus.getStatus(), expectedStatus.getStatus());
     }
 
+    @Test
+    public void testExecuteFailure() throws Exception {
+        // Setup
+        ActionType actionType = ActionType.PRE_ISSUE_ACCESS_TOKEN;
+        Map<String, Object> eventContext = Collections.emptyMap();
+
+        // Mock Action and its dependencies
+        Action action = createAction();
+
+        // Mock ActionManagementService
+        when(actionManagementService.getActionsByActionType(any(), any())).thenReturn(
+                Collections.singletonList(action));
+
+        // Mock ActionRequestBuilder and ActionResponseProcessor
+        actionExecutionRequestBuilderFactory.when(
+                        () -> ActionExecutionRequestBuilderFactory.getActionExecutionRequestBuilder(any()))
+                .thenReturn(actionExecutionRequestBuilder);
+        actionExecutionResponseProcessorFactory.when(() -> ActionExecutionResponseProcessorFactory
+                        .getActionExecutionResponseProcessor(any()))
+                .thenReturn(actionExecutionResponseProcessor);
+
+        // Configure request builder
+        when(actionExecutionRequestBuilder.getSupportedActionType()).thenReturn(actionType);
+        when(actionExecutionRequestBuilder.buildActionExecutionRequest(eventContext)).thenReturn(
+                mock(ActionExecutionRequest.class));
+
+        // Mock APIClient response
+        ActionInvocationResponse actionInvocationResponse = createFailureActionInvocationResponse();
+        when(apiClient.callAPI(any(), any(), any())).thenReturn(actionInvocationResponse);
+
+        // Configure response processor
+        ActionExecutionStatus expectedStatus =
+                new ActionExecutionStatus(ActionExecutionStatus.Status.FAILED, eventContext);
+        when(actionExecutionResponseProcessor.getSupportedActionType()).thenReturn(actionType);
+        when(actionExecutionResponseProcessor.processFailureResponse(any(), any(), any())).thenReturn(
+                expectedStatus);
+
+        // Execute and assert
+        ActionExecutionStatus actualStatus =
+                actionExecutorService.execute(actionType, eventContext, "tenantDomain");
+        assertEquals(actualStatus.getStatus(), expectedStatus.getStatus());
```

Comment:
Assert for failure message which is expected as required

Reply:
This is ActionExecution assert and failure message need to be handle from the corresponding ActionExecutionResponseProcessor. 
The null check for the failure message in the Action Invocation Response is done by the ActionInvocationFailureResponse class.